### PR TITLE
Block CAP-affected users from adding personal email logins

### PR DESCRIPTION
**Pull request by carl-codeorg**
This solves a few issues relating to LTI accounts and accounts missing `us_state` and/or `age`:
- Disables the "add a password" (aka personal login) form on the account settings page, regardless of account type, if the user would otherwise be locked out of adding oauth-based logins. The criteria being that they're either U13 and in a CAP state without permission, or missing age, state, or country code.
- Adjusts the criteria for being locked out of adding a new login via the Managed Linked Accounts table to include age as a required field.
- Refactors the `edit` route in the `registrations_controller` to reduce confusion caused by all the similarly-named lockout-related variables.

Screenshot 2024-10-17 at 4 05 31 PM

## Links

[JIRA](https://codedotorg.atlassian.net/browse/P20-1236)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Review by nicklathe — APPROVED** [review]
LGTM

File: dashboard/app/views/devise/registrations/edit.html.haml [hunk 1]
```diff
@@ -72,7 +72,7 @@
   = form_for(resource, as: resource_name, url: registration_path(resource_name), html: {method: :put, id: 'lti-sync-settings-form'}) do |f|
     = f.hidden_field :lti_roster_sync_enabled
 
-- if @potentially_locked && experiment_value('cpa-partial-lockout', request)
+- if (!@personal_account_linking_enabled || Policies::ChildAccount::ComplianceState.permission_granted?(current_user)) && experiment_value('cpa-partial-lockout', request)
```
Comment (daynew) [inline dashboard/app/views/devise/registrations/edit.html.haml:75]:
```suggestion
- if !@personal_account_linking_enabled && experiment_value('cpa-partial-lockout', request)
```
Do we need this `permission_granted?` given that `can_link_new_personal_account?` already calls it?

File: dashboard/lib/policies/child_account.rb [hunk 2]
```diff
@@ -196,13 +196,17 @@ def self.underage?(user)
...
-    return false unless user.us_state && user.country_code
-    return true unless user.birthday
+    return false unless has_required_information?(user)
+    return true unless ['US', 'RD'].include?(user.country_code)
     return true unless underage?(user)
 
     ComplianceState.permission_granted?(user)
   end
 
+  def self.has_required_information?(user)
```
Comment (daynew) [inline dashboard/lib/policies/child_account.rb:207]:
```suggestion
  # Returns true if the user has provided the minimum information we need to decide if their account is affected by our Child Account Policy.
  def self.has_required_information?(user)
```

**Review by daynew — APPROVED** [review]

File: dashboard/app/views/devise/registrations/edit.html.haml (hunk 1 above)
Comment (carl-codeorg) [reply dashboard/app/views/devise/registrations/edit.html.haml:75]: We do actually need this - the distinction is that it's checking for permission being granted even if personal account linking is also enabled. This lets us display the "granted" status even after the account is fully unlocked.

**Pull request merged**